docs/content/preview/architecture/transactions/read-restart-error.md

```diff
@@ -32,6 +32,12 @@ A detailed scenario that explains how clock skew can result in the above mention
 
   * It should return the data even if the client issued the read from a different node after writing the data, because the following guarantee needs to be maintained: the database should always return data that was committed in the past (past refers to the user-perceived past, and not based on machine clocks).
 
-  * It should not return the data if the write is performed in the future (that is, after the read point had been chosen) **and** had a write timestamp later than the read point. Because if that is allowed, everything written in future would be trivially visible to the read. Note that the latter condition is important because it is okay to return data that was written after the read point was picked, if the write timestamp was earlier than the read point (this doesn't break and consistency or isolation guarantees).
+  * It should not return the data if the write is performed in the future (that is, after the read point had been chosen) **and** had a write timestamp later than the read point. Because if that is allowed, everything written in future would be trivially visible to the read. Note that the latter condition is important because it is okay to return data that was written after the read point was picked, if the write timestamp was earlier than the read point (this doesn't break any consistency or isolation guarantees).
 
 * If node `N2` finds writes in the range `(T1, T1+max_clock_skew]`, to avoid breaking the strong guarantee that a reader should always be able to read what was committed earlier, and to avoid reading data with a later write timestamp which was also actually written after the read had been issued, node `N2` raises a `Read restart` error.
+
+## Troubleshooting
+
+1. Implement retry logic in the application. Application retries can help with mitigating read restart errors. Moreover, there are other ways in which a statement or a transaction may fail such as transaction conflicts or infrastructure failures. Therefore, a retry mechanism is strongly recommended for a cloud-native, distributed database such as YugabyteDB.
+2. Use SERIALIZABLE, READ ONLY, DEFERRABLE mode whenever possible. Read restart errors usually occur when the query is a SELECT statement with a large output footprint. These are better run with the above-mentioned transaction characteristics anyway. Using DEFERRABLE will avoid a read restart error altogether. However, the tradeoff is that the statement waits out the clock skew before reading the data. This is not an issue for large SELECT statements. Tip: use `SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;` to change the default transaction properties.
```

Comment:
```suggestion
- Use SERIALIZABLE, READ ONLY, or DEFERRABLE mode whenever possible. Read restart errors usually occur when the query is a SELECT statement with a large output footprint. These are better run with the above-mentioned transaction characteristics anyway.

  Using DEFERRABLE will avoid a read restart error altogether. However, the tradeoff is that the statement waits out the clock skew before reading the data. This is not an issue for large SELECT statements.
  
  Tip: use `SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;` to change the default transaction properties.
```

Reply:
.... Read restart errors usually occur when the query is a SELECT statement with a large output footprint and there are concurrent writes that satisfy the SELECT statement.


"These are better run with the above-mentioned transaction characteristics anyway." -- I am not sure we should say this because if customers start using this, they won't go back once we make big improvements to reducing read restart errors? Also, there is https://github.com/yugabyte/yugabyte-db/issues/23120. On a side note, I realize that we need to prioritize this issue too.

However, the tradeoff is that the statement waits out the maximum permissible clock skew before reading the data (which is ysql_max_clock_skew_usecs that has a default of 500ms?)

.... This is not an issue for large SELECT statements because of the large latency of the read?

Reply:
Not including "or" since this is PostgreSQL specific syntax and it does not have that.